### PR TITLE
[Merged by Bors] - feat: a family of compatible triangularisable endomorphisms are simultaneously triangularisable

### PR DESCRIPTION
**Pull request by ocfnash**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by j-loreaux — COMMENTED** [review]
There's a bit of mess created from using `Set.MapsTo` as the condition for invariant subspaces. Indeed, you really *want* it to unfold to `∀ x ∈ p, f x ∈ p`, but it actually unfolds to `∀ x ∈ ↑p, f x ∈ ↑p`, where `↑` is the coercion to `Set`. You can see the annoyance this causes in `mapsTo_restrict_maxGenEigenspace_restrict_of_mapsTo` especially. I'm not going to hold up this PR over it, but I think there are some discrepancies with the way we handle invariant subspaces throughout the library (e.g., `LinearMap.restrict` itself uses `hf : ∀ x ∈ p, f x ∈ q`). I think we should create a bespoke `Submodule.Invariant` predicate that has the right defeqs, and the things to go between that and `Set.MapsTo`.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 1]
```diff
@@ -449,6 +449,44 @@ lemma _root_.Submodule.inf_genEigenspace (f : End R M) (p : Submodule R M) {k :
...
+  rfl
+
+lemma mapsTo_restrict_maxGenEigenspace_restrict_of_mapsTo
+    {p : Submodule R M} (f g : End R M) (hf : MapsTo f p p) (hg : MapsTo g p p) {μ : R}
+    (h : MapsTo f (g.maxGenEigenspace μ) (g.maxGenEigenspace μ)) :
+    MapsTo (f.restrict hf)
+      (maxGenEigenspace (g.restrict hg) μ)
+      (maxGenEigenspace (g.restrict hg) μ) := by
+  intro ⟨x, hx⟩ hx'
+  have : μ • (1 : Module.End R p) = (μ • 1 : Module.End R M).restrict (fun _ ↦ p.smul_mem μ) := rfl
```
[the remark this answers is not in the document]
Comment (ocfnash) [reply Mathlib/LinearAlgebra/Eigenspace/Basic.lean:484]: Done.

**Comment by ocfnash** [comment]
> I think we should create a bespoke `Submodule.Invariant` predicate that has the right defeqs, and the things to go between that and `Set.MapsTo`.

Agreed.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 2]
```diff
@@ -449,6 +449,43 @@ lemma _root_.Submodule.inf_genEigenspace (f : End R M) (p : Submodule R M) {k :
       (genEigenspace (LinearMap.restrict f hfp) μ k).map p.subtype := by
   rw [f.genEigenspace_restrict _ _ _ hfp, Submodule.map_comap_eq, Submodule.range_subtype]
 
+lemma _root_.Submodule.inf_iInf_maxGenEigenspace_of_forall_mapsTo {ι : Type*} {μ : ι → R}
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:452]: This one

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 3]
```diff
@@ -449,6 +449,43 @@ lemma _root_.Submodule.inf_genEigenspace (f : End R M) (p : Submodule R M) {k :
...
+    (f : ι → End R M) (p : Submodule R M) (hfp : ∀ i, MapsTo (f i) p p) :
+    p ⊓ ⨅ i, (f i).maxGenEigenspace (μ i) =
+      (⨅ i, maxGenEigenspace ((f i).restrict (hfp i)) (μ i)).map p.subtype := by
+  cases isEmpty_or_nonempty ι
+  · simp [iInf_of_isEmpty]
+  · simp_rw [inf_iInf, maxGenEigenspace, ((f _).genEigenspace _).mono.directed_le.inf_iSup_eq,
+      p.inf_genEigenspace _ (hfp _), ← Submodule.map_iSup, Submodule.map_iInf _ p.injective_subtype]
+
+lemma iInf_maxGenEigenspace_restrict_map_subtype_eq
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:461]: and this

File: Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean [hunk 4]
```diff
@@ -222,3 +222,48 @@ theorem Module.End.iSup_genEigenspace_restrict_eq_top
   simp_rw [Submodule.inf_genEigenspace f p h, Submodule.comap_subtype_self,
     ← Submodule.map_iSup, Submodule.comap_map_eq_of_injective h_inj] at this
   exact this.symm
+
+lemma Module.End.iSup_iInf_maxGenEigenspace_eq_top_of_commute {ι : Type*} [FiniteDimensional K V]
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean:226]: and this

**Review by j-loreaux — APPROVED** [review]
The statements of these lemmas are sufficiently hard to parse at a glance that it would be nice to have docstrings. I'll mark the ones I mean. Other than that:

bors d+

**Comment by ocfnash** [comment]
Thanks for the review!

**Comment by ocfnash** [comment]
bors merge
